File: sdk/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/InProcessTransformResult.java
/*
 * Copyright (C) 2015 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */
package com.google.cloud.dataflow.sdk.runners.inprocess;

import com.google.cloud.dataflow.sdk.runners.inprocess.InMemoryWatermarkManager.TimerUpdate;
import com.google.cloud.dataflow.sdk.runners.inprocess.InProcessPipelineRunner.UncommittedBundle;
import com.google.cloud.dataflow.sdk.transforms.AppliedPTransform;
import com.google.cloud.dataflow.sdk.transforms.PTransform;
import com.google.cloud.dataflow.sdk.transforms.windowing.BoundedWindow;
import com.google.cloud.dataflow.sdk.util.common.CounterSet;
import com.google.cloud.dataflow.sdk.util.state.CopyOnAccessInMemoryStateInternals;

import org.joda.time.Instant;

import javax.annotation.Nullable;

/**
 * The result of evaluating an {@link AppliedPTransform} with a {@link TransformEvaluator}.
 */
public interface InProcessTransformResult {
  /**
   * Returns the {@link AppliedPTransform} that produced this result.
   */
  AppliedPTransform<?, ?, ?> getTransform();

  /**
   * Returns the {@link UncommittedBundle (uncommitted) Bundles} output by this transform. These
   * will be committed by the evaluation context as part of completing this result.
   */
  Iterable<? extends UncommittedBundle<?>> getOutputBundles();

  /**
   * Returns the {@link CounterSet} used by this {@link PTransform}, or null if this transform did
   * not use a {@link CounterSet}.
   */
  @Nullable CounterSet getCounters();

  /**
   * Returns the Watermark Hold for the transform at the time this result was produced.
   *
   * If the transform does not set any watermark hold, returns
   * {@link BoundedWindow#TIMESTAMP_MAX_VALUE}.
   */
  Instant getWatermarkHold();

  /**
   * Returns the State used by the transform.
   *
   * If this evaluation did not access state, this may return null.
   */
  CopyOnAccessInMemoryStateInternals<?> getState();

  /**
   * Returns a TimerUpdateBuilder that was produced as a result of this evaluation. If the
   * evaluation was triggered due to the delivery of one or more timers, those timers must be added
   * to the builder before it is complete.
   *
   * <p>If this evaluation did not add or remove any timers, returns an empty TimerUpdate.
   */
  TimerUpdate getTimerUpdate();

}

File: sdks/python/apache_beam/runners/inprocess/inprocess_transform_result.py
#
# Licensed to the Apache Software Foundation (ASF) under one or more
# contributor license agreements.  See the NOTICE file distributed with
# this work for additional information regarding copyright ownership.
# The ASF licenses this file to You under the Apache License, Version 2.0
# (the "License"); you may not use this file except in compliance with
# the License.  You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

"""The result of evaluating an AppliedPTransform with a TransformEvaluator."""

from __future__ import absolute_import


class InProcessTransformResult(object):
  """The result of evaluating an AppliedPTransform with a TransformEvaluator."""

  def __init__(self, applied_ptransform, uncommitted_output_bundles, state,
               timer_update, counters, watermark_hold,
               undeclared_tag_values=None):
    self._applied_ptransform = applied_ptransform
    self._uncommitted_output_bundles = uncommitted_output_bundles
    self._state = state
    self._timer_update = timer_update
    self._counters = counters
    self._watermark_hold = watermark_hold
    # Only used when caching (materializing) all values is requested.
    self._undeclared_tag_values = undeclared_tag_values

  @property
  def transform(self):
    return self._applied_ptransform

  @property
  def output_bundles(self):
    return self._uncommitted_output_bundles

  @property
  def state(self):
    return self._state

  @property
  def counters(self):
    return self._counters

  @property
  def watermark_hold(self):
    return self._watermark_hold

  @property
  def undeclared_tag_values(self):
    return self._undeclared_tag_values
